### PR TITLE
[YSQL] Added functional spec for pessimistic locking in architecture/design/

### PR DESCRIPTION
**Pull request by pkj415**

### CODE REVIEW

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 1]
```diff
@@ -0,0 +1,661 @@
...
+
+<span style="text-decoration:underline;">PostgreSQL supports 4 types of row-level locks -</span>
+
+
+
+1. Exclusive modes (i.e., only 1 txn can hold such a lock at any time) -
+    1. `FOR UPDATE` - takes exclusive lock on the whole row (blocks the shared locks as well)
+    2. `FOR NO KEY UPDATE` - takes exclusive lock on row but allows share lock on pk i.e., `FOR KEY SHARE`
+2. Share access modes
+    3. `FOR SHARE` - shared lock on full row i.e., blocks all exclusive locks
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:16]: Restart numbering. Ideally, just do all numbers as `1.` and let markdown figure out the sequencing. Indentation will sort out the sub-lists.

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 2]
```diff
@@ -0,0 +1,661 @@
...
+
+## 1 Introduction
+
+To understand pessimistic locking behaviour in Pg, it is important to take note of some points about row-level locking and its interaction with DMLs.
+
+<span style="text-decoration:underline;">PostgreSQL supports 4 types of row-level locks -</span>
+
+
+
+1. Exclusive modes (i.e., only 1 txn can hold such a lock at any time) -
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:12]: Use the full terms:
"transaction"
"primary key"
"PostgreSQL"
etc.

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 3]
```diff
@@ -0,0 +1,661 @@
...
+
+
+2. **Lock-lock conflict:** this conflict only refers to one lock blocking another i.e., just waiting behaviour. It **doesn’t** refer to a serialization error. In other words, two conflicting locks can only result in blocking and not in a serialization error (or “write” conflict as we roughly call it). Also note that these locks can be taken explicitly using a `SELECT FOR `or implicitly using a DML. So there can be 4 cases - implicit-implicit, implicit-explicit, explicit-implicit, explicit-explicit lock-lock conflicts.
+
+    _(refer example 1 for explicit lock-explicit lock conflict)._
+
+
+DMLs implicitly take the following row-level locks:
+
+
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:44]: Everywhere: more than one blank line gets ignored by the renderer

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 4]
```diff
@@ -0,0 +1,661 @@
...
+2. **Txn already in queue might jump older txns in the queue: **Multiple transactions might be blocked to lock (either implicitly via write or explicitly) the same tuple and this results in a queue. When an active txn ends, txns that entered the queue earlier are checked first for unblocking. Txns that don’t conflict with the active txns are unblocked to acquire locks and become active. Note that this implies that a transaction that conflicts with earlier transactions blocked in the queue might still be unblocked if it doesn’t conflict with active transactions (which could lead to starvation).
+3. In case <code>[statement_timeout or lock_timeout](https://www.postgresql.org/docs/13/runtime-config-client.html)</code> is non-zero, a blocked txn will abort after the timeout. (example 4). A zero timeout would imply indefinite waiting.
+
+NOTE: Refer this [link](https://postgrespro.com/blog/pgsql/5968005) for a good overview of row-level locking.
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:74]: This article, this blog post... anything but link

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 5]
```diff
@@ -0,0 +1,661 @@
...
+
+
+
+1. Firstly, support lock-lock type of conflict that only blocks, in YSQL. This is to split the current notion of conflict in YSQL into the two finer notions that Postgres supports.
+2. In case a transaction T1 tries to acquire a lock as part of a implicit/explicit lock acquisition that conflicts with existing locks held by active txns -
+    1. It has to wait for all txns that have conflicting locks to end (waiting behaviour)
+    2. Throw a serialization error only if a modification that conflicts with the lock has occurred.
+    3. Ability to timeout and remove self from queue
+3. A [distributed deadlock detection algorithm](https://docs.google.com/document/d/1E4LHGmVZuTlr36_uczPuE6aAjoLT4sfr1o0YCtzUUPc/edit#) to break cycles. The following properties are required -
+    4. No false positives
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:87]: Restart numbering

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 6]
```diff
@@ -0,0 +1,661 @@
...
+    5. Bound on latency = O (cycle size)
+4. Add a metric to measure the “intensity” of starvation by measuring the number instances where a transaction jumps the wait queue ahead of other waiting transactions that it conflicts with just because it doesn’t conflict with active transactions.
+
+Once pessimistic locking is supported, YSQL will have the same behaviour as Postgres with regards to transaction waiting behaviour in case of writing/locking, with **_two exceptions_**:
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:91]: Bold or italic

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 7]
```diff
@@ -0,0 +1,661 @@
...
+
+
+
+1. For Serializable isolation level, there will be an extra YSQL-only behaviour of waiting in case a read (or write) intent is to be written that conflicts with an already existing write (or read) intent from another transaction. In Pg’s Serializable isolation implementation (i.e., SSI), reads don’t take implicit locks and hence don’t “lock” conflict with other writes. And hence, a read doesn’t wait on another write (and vice versa) in Pg’s Serializable isolation.
+2. YSQL writes fine grained column level intents in case of modifications to specific columns only. This will allow us to be more fine grained so that modifications to different columns of a row need not result in waiting (possibly followed by a serialization error). This is one aspect in which YSQL would turn out to be **_better than_** Pg - and semantically different in a hopefully beneficial way.
+    1. In case customers are relying on this more coarse-grained locking behavior of PostgreSQL, we can’t help them without significant overhaul of existing behavior
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:97]: Is this indented on purpose? It's odd to have a single sub-item

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 8]
```diff
@@ -0,0 +1,661 @@
...
+
+
+## 3 Surface Area & Usage
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:100]:
```suggestion
## 3 Surface Area and Usage
```

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 9]
```diff
@@ -0,0 +1,661 @@
...
+
+#### Surface Area
+
+**yb_use_pessimistic_locking**: a cluster level gflag to turn on pessimistic locking. This will require a cluster restart. Note that all txns in the cluster will either use pessimistic locking or optimistic locking.
+
+
+#### Expected Behavior
+
+
+```
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:111]:
```suggestion
```sql
```

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 10]
```diff
@@ -0,0 +1,661 @@
...
+
+
+
+1. **Explicit lock - Explicit lock interaction (only blocks)**
+
+<table>
+  <tr>
+   <td>
+<code>> begin transaction isolation level repeatable read; \
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:123]: This seems like it won't render right with the two angle brackets...

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 11]
```diff
@@ -0,0 +1,661 @@
...
+  </tr>
+</table>
+
+
+
+
+2. READ COMMITTED isolation level (exactly as in Pg) has a dependency on pessimistic locking. Note that in general pessimistic locking is orthogonal to isolation levels but READ COMMITTED specifically has a “dependency” on pessimistic locking. To be precise, on facing a conflict, a txn has to wait for the conflicting txn to rollback/commit. On commit, the txn will fetch the latest version of the row and work on that.
+
+
+#### _Versioning & upgrades_
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:650]:
```suggestion
#### Versioning and upgrades
```

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 12]
```diff
@@ -0,0 +1,661 @@
...
+
+
+## 4 Compatibility
+
+
+#### Cross feature interaction
+
+
+
+1. Interaction with savepoints -
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:574]:
```suggestion
```

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 13]
```diff
@@ -0,0 +1,661 @@
...
+This feature is upgrade and downgrade safe. When turning the gflag on/off, or during rolling restarts across versions with the flag “on” in the higher version, if some nodes have pessimistic locking on and some don’t, users will experience mixed (but still correct) behavior. A mix of both optimistic locking and pessimistic locking will result in the following additional YSQL specific semantics -
+
+
+
+1. Assume T1 follows optimistic locking and T2 follows pessimistic locking
+2. If a txn T1 uses optimistic locking and sees txns that have written conflicting intents -
+    1. if there is even 1 conflicting txn that follows pessimistic locking, abort T1 
+    2. Else, behaviour is as today
+3. If a txn T1 uses pessimistic locking and sees txns that have written conflicting intents -
+    3. Wait for all conflicting txns to end
```
Comment (polarweasel) [inline architecture/design/pessimistic-locking-functional-spec.md:661]:
```suggestion
    1. Wait for all conflicting txns to end
```

**Review by polarweasel — CHANGES_REQUESTED** [review]
Lots of changes noted...

File: architecture/design/pessimistic-locking-functional-spec.md (hunk 2 above)
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:12]: done

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 14]
```diff
@@ -0,0 +1,661 @@
...
+
+To understand pessimistic locking behaviour in Pg, it is important to take note of some points about row-level locking and its interaction with DMLs.
+
+<span style="text-decoration:underline;">PostgreSQL supports 4 types of row-level locks -</span>
+
+
+
+1. Exclusive modes (i.e., only 1 txn can hold such a lock at any time) -
+    1. `FOR UPDATE` - takes exclusive lock on the whole row (blocks the shared locks as well)
+    2. `FOR NO KEY UPDATE` - takes exclusive lock on row but allows share lock on pk i.e., `FOR KEY SHARE`
```
[the remark this answers is not in the document]
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:14]: done

File: architecture/design/pessimistic-locking-functional-spec.md (hunk 1 above)
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:16]: done

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 15]
```diff
@@ -0,0 +1,661 @@
...
+2. Share access modes
+    3. `FOR SHARE` - shared lock on full row i.e., blocks all exclusive locks
+    4. `FOR KEY SHARE` - shared lock on key i.e., allows `FOR NO KEY UPDATE`
+
+<span style="text-decoration:underline;">PostgreSQL has two notions of conflict -</span>
+
+
+
+1. **Lock-modification conflict: **(modification is a change in tuple via a DML) _(refer example 2)._
```
[the remark this answers is not in the document]
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:23]: done

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 16]
```diff
@@ -0,0 +1,661 @@
...
+    1. `FOR UPDATE`: doesn’t allow any concurrent modification to the locked tuple.
+    2. `FOR NO KEY UPDATE`: doesn’t allow any concurrent modification to the locked tuple.
+    3. `FOR SHARE`: doesn’t allow any concurrent modification to the locked tuple.
+    4. `FOR KEY SHARE`: allow only modification to non-pk cols
+
+Lock-modification conflict only happens when the period from lock acquire to lock release intersects/ overlaps with the read to commit time of the transaction performing the modification.
+
+Lock-modification conflicts result in serialization errors. Note that the lock might be issued before or after the modification (refer example 2i and 2ii).
```
[the remark this answers is not in the document]
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:31]: done

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 17]
```diff
@@ -0,0 +1,661 @@
...
+
+Note that in READ COMMITTED isolation, when a lock-modification conflict occurs, Pg does some post-processing (called “READ COMMITTED update checking steps”) instead of throwing a 40001.
+
+
+
+2. **Lock-lock conflict:** this conflict only refers to one lock blocking another i.e., just waiting behaviour. It **doesn’t** refer to a serialization error. In other words, two conflicting locks can only result in blocking and not in a serialization error (or “write” conflict as we roughly call it). Also note that these locks can be taken explicitly using a `SELECT FOR `or implicitly using a DML. So there can be 4 cases - implicit-implicit, implicit-explicit, explicit-implicit, explicit-explicit lock-lock conflicts.
```
[the remark this answers is not in the document]
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:37]: done

File: architecture/design/pessimistic-locking-functional-spec.md (hunk 3 above)
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:44]: makes sense

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 18]
```diff
@@ -0,0 +1,661 @@
...
+
+DMLs implicitly take the following row-level locks:
+
+
+
+1. `DELETE` -> `FOR UPDATE`
+2. `UPDATE` -> `FOR UPDATE` if updating a column which has a unique index on it (one that can be used in a foreign key)
+3. All `UPDATEs` other than those in point 2 -> `FOR NO KEY UPDATE`.
+
+To make it easier to think, “DML” = “locking” + “modification” of a tuple.
```
[the remark this answers is not in the document]
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:50]: nice one!

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 19]
```diff
@@ -0,0 +1,661 @@
...
+
+Given that writing = locking + modification, the definition of lock-modification conflict is more generic to think of for serialization errors (the usual write-write “conflict” we talk about is just a special case of this.** A write-write conflict = implicit lock - modification conflict**).
```
[the remark this answers is not in the document]
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:52]: done

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 20]
```diff
@@ -0,0 +1,661 @@
...
+
+<span style="text-decoration:underline;">Conflicts in YSQL -</span>
```
[the remark this answers is not in the document]
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:54]: done

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 21]
```diff
@@ -0,0 +1,661 @@
...
+
+YSQL does not differentiate between a lock-lock conflict and a lock-modification conflict. In either case, we will trigger conflict resolution. Conflict resolution in YSQL doesn’t lead to any blocking, but would surely abort either itself or all conflicting transactions based on priorities of the transactions (which are randomly chosen). This is _optimistic locking_.
+
+Moreover, unlike Pg’s Serializable isolation, in YSQL’s Serializable isolation, plain SELECTs without explicit lock also implicitly take _FOR SHARE_ lock. And this can result in a plain read DML to conflict with a write DML (which will result in triggering conflict resolution).
+
+<span style="text-decoration:underline;">Pessimistic Locking in PostgreSQL -</span>
+
+PostgreSQL uses **_pessimistic locking_**, where a transaction will **wait upon encountering a lock-lock conflict** in order to acquire the locks it needs and proceed.
```
[the remark this answers is not in the document]
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:62]: done

File: architecture/design/pessimistic-locking-functional-spec.md (hunk 4 above)
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:74]: done

File: architecture/design/pessimistic-locking-functional-spec.md (hunk 5 above)
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:87]: done

File: architecture/design/pessimistic-locking-functional-spec.md (hunk 6 above)
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:91]: done

File: architecture/design/pessimistic-locking-functional-spec.md (hunk 8 above)
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:100]: done

File: architecture/design/pessimistic-locking-functional-spec.md (hunk 9 above)
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:111]: done

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 22]
```diff
@@ -0,0 +1,661 @@
...
+```
+create table test (k int primary key, v int);
+insert into test values (1, 1);
+```
+
+
+
+1. **Explicit lock - Explicit lock interaction (only blocks)**
+
+<table>
```
[the remark this answers is not in the document]
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:120]: okay, i will get in touch with you on how to do this in the next iteration

File: architecture/design/pessimistic-locking-functional-spec.md (hunk 12 above)
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:574]: done

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 23]
```diff
@@ -0,0 +1,661 @@
...
+
+## 4 Compatibility
+
+
+#### Cross feature interaction
+
+
+
+1. Interaction with savepoints -
+    1. Rollback a savepoint might result in release of a lock and hence resolve some lock-lock waits.
```
[the remark this answers is not in the document]
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:575]: done

File: architecture/design/pessimistic-locking-functional-spec.md [hunk 24]
```diff
@@ -0,0 +1,661 @@
...
+   </td>
+   <td><code>> commit;</code>
+   </td>
+  </tr>
+</table>
+
+
+
+
+2. READ COMMITTED isolation level (exactly as in Pg) has a dependency on pessimistic locking. Note that in general pessimistic locking is orthogonal to isolation levels but READ COMMITTED specifically has a “dependency” on pessimistic locking. To be precise, on facing a conflict, a txn has to wait for the conflicting txn to rollback/commit. On commit, the txn will fetch the latest version of the row and work on that.
```
[the remark this answers is not in the document]
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:647]: done

File: architecture/design/pessimistic-locking-functional-spec.md (hunk 11 above)
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:650]: done

File: architecture/design/pessimistic-locking-functional-spec.md (hunk 10 above)
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md:123]: done

File: architecture/design/pessimistic-locking-functional-spec.md (hunk 13 above)
Comment (pkj415) [reply architecture/design/pessimistic-locking-functional-spec.md]: done

**Comment by pkj415** [comment]
Merging this post discussion with @polarweasel . He is out for sometime and this needs to land urgently.

**Pull request merged**